File: java/src/test/java/it/giannibombelli/wsc2026/payment/domain/payment/TransactionTest.java
package it.giannibombelli.wsc2026.payment.domain.payment;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class TransactionTest {

    @Nested
    class Start {
        @Test
        void shouldCreateStartedTransaction() {
            TransactionId transactionId = EntityId.generate(TransactionId::new);
            Provider provider = Provider.GIFT_CARD;
            ProviderReference providerReference = new ProviderReference(UUID.randomUUID());
            Money amount = new Money(BigDecimal.TEN);
            Timestamp startedAt = new Timestamp(Instant.parse("2026-06-07T10:00:00Z"));

            Transaction transaction = Transaction.start(transactionId, provider, providerReference, amount, startedAt);

            assertThat(transaction.id()).isEqualTo(transactionId);
            assertThat(transaction.provider()).isEqualTo(provider);
            assertThat(transaction.providerReference()).isEqualTo(providerReference);
            assertThat(transaction.amount()).isEqualTo(amount);
            assertThat(transaction.status()).isEqualTo(TransactionStatus.STARTED);
            assertThat(transaction.startedAt()).isEqualTo(startedAt);
            assertThat(transaction.completedAt()).isNull();
        }

        @Test
        void shouldFailIfParametersAreNull() {
            TransactionId transactionId = EntityId.generate(TransactionId::new);
            Provider provider = Provider.GIFT_CARD;
            ProviderReference providerReference = new ProviderReference(UUID.randomUUID());
            Money amount = new Money(BigDecimal.TEN);
            Timestamp startedAt = new Timestamp(Instant.parse("2026-06-07T10:00:00Z"));

            assertThatThrownBy(() -> Transaction.start(null, provider, providerReference, amount, startedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> Transaction.start(transactionId, null, providerReference, amount, startedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> Transaction.start(transactionId, provider, providerReference, null, startedAt))
                .isInstanceOf(IllegalArgumentException.class);

            assertThatThrownBy(() -> Transaction.start(transactionId, provider, providerReference, amount, null))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }

    @Nested
    class Accept {
        @Test
        void shouldFailIfCompletedAtIsNull() {
            Transaction transaction = createStartedTransaction();

            assertThatThrownBy(() -> transaction.accept((Timestamp) null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldFailIfTransactionIsNotStarted() {
            Transaction accepted = createStartedTransaction().accept(Timestamp.now());

            assertThatThrownBy(() -> accepted.accept(Timestamp.now()))
                .isInstanceOf(IllegalStateException.class);
        }
    }

    @Nested
    class Reject {
        @Test
        void shouldFailIfCompletedAtIsNull() {
            Transaction transaction = createStartedTransaction();

            assertThatThrownBy(() -> transaction.reject((Timestamp) null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldFailIfTransactionIsNotStarted() {
            Transaction rejected = createStartedTransaction().reject(Timestamp.now());

            assertThatThrownBy(() -> rejected.reject(Timestamp.now()))
                .isInstanceOf(IllegalStateException.class);
        }
    }

    private Transaction createStartedTransaction() {
        return Transaction.start(
            EntityId.generate(TransactionId::new),
            Provider.GIFT_CARD,
            new ProviderReference(UUID.randomUUID()),
            new Money(BigDecimal.TEN),
            new Timestamp(Instant.parse("2026-06-07T10:00:00Z"))
        );
    }
}

File: typescript/tests/payment/domain/payment/transaction.test.ts
import { describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { Provider } from '@/payment/domain/payment/provider.js';
import { ProviderReference } from '@/payment/domain/payment/providerReference.js';
import { Transaction } from '@/payment/domain/payment/transaction.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { TransactionId } from '@/payment/domain/payment/transactionId.js';
import { TransactionStatus } from '@/payment/domain/payment/transactionStatus.js';

describe('Transaction', () => {
  describe('start', () => {
    it('should create started transaction', () => {
      const transactionId = generateId((value) => new TransactionId(value));
      const provider = Provider.GIFT_CARD;
      const providerReference = new ProviderReference(Uuid.generate());
      const amount = new Money(10);
      const startedAt = new Timestamp(new Date('2026-06-07T10:00:00.000Z'));

      const transaction = Transaction.start(transactionId, provider, providerReference, amount, startedAt);

      expect(transaction.id()).toEqual(transactionId);
      expect(transaction.provider()).toEqual(provider);
      expect(transaction.providerReference()).toEqual(providerReference);
      expect(transaction.amount()).toEqual(amount);
      expect(transaction.status()).toEqual(TransactionStatus.STARTED);
      expect(transaction.startedAt()).toEqual(startedAt);
      expect(transaction.completedAt()).toBeNull();
    });

    it('should fail if parameters are null', () => {
      const transactionId = generateId((value) => new TransactionId(value));
      const provider = Provider.GIFT_CARD;
      const providerReference = new ProviderReference(Uuid.generate());
      const amount = new Money(10);
      const startedAt = new Timestamp(new Date('2026-06-07T10:00:00.000Z'));

      expect(() => Transaction.start(null as unknown as TransactionId, provider, providerReference, amount, startedAt)).toThrow();
      expect(() => Transaction.start(transactionId, null as unknown as Provider, providerReference, amount, startedAt)).toThrow();
      expect(() => Transaction.start(transactionId, provider, providerReference, null as unknown as Money, startedAt)).toThrow();
      expect(() => Transaction.start(transactionId, provider, providerReference, amount, null as unknown as Timestamp)).toThrow();
    });
  });

  describe('accept', () => {
    it('should fail if completed at is null', () => {
      const transaction = createStartedTransaction();

      expect(() => transaction.accept(null as unknown as Timestamp)).toThrow();
    });

    it('should fail if transaction is not started', () => {
      const accepted = createStartedTransaction().accept(Timestamp.now());

      expect(() => accepted.accept(Timestamp.now())).toThrow();
    });
  });

  describe('reject', () => {
    it('should fail if completed at is null', () => {
      const transaction = createStartedTransaction();

      expect(() => transaction.reject(null as unknown as Timestamp)).toThrow();
    });

    it('should fail if transaction is not started', () => {
      const rejected = createStartedTransaction().reject(Timestamp.now());

      expect(() => rejected.reject(Timestamp.now())).toThrow();
    });
  });

  function createStartedTransaction(): Transaction {
    return Transaction.start(
      generateId((value) => new TransactionId(value)),
      Provider.GIFT_CARD,
      new ProviderReference(Uuid.generate()),
      new Money(10),
      new Timestamp(new Date('2026-06-07T10:00:00.000Z'))
    );
  }
});
